packages/react-router-dom/index.tsx

```diff
@@ -470,48 +700,312 @@ export function useSearchParams(defaultInit?: URLSearchParamsInit) {
   return [searchParams, setSearchParams] as const;
 }
 
-export type ParamKeyValuePair = [string, string];
+/**
+ * Submits a HTML `<form>` to the server without reloading the page.
+ */
+export interface SubmitFunction {
+  (
+    /**
+     * Specifies the `<form>` to be submitted to the server, a specific
+     * `<button>` or `<input type="submit">` to use to submit the form, or some
+     * arbitrary data to submit.
+     *
+     * Note: When using a `<button>` its `name` and `value` will also be
+     * included in the form data that is submitted.
+     */
+    target:
+      | HTMLFormElement
+      | HTMLButtonElement
+      | HTMLInputElement
+      | FormData
+      | URLSearchParams
+      | { [name: string]: string }
+      | null,
+
+    /**
+     * Options that override the `<form>`'s own attributes. Required when
+     * submitting arbitrary data without a backing `<form>`.
+     */
+    options?: SubmitOptions
+  ): void;
+}
+
+/**
+ * Returns a function that may be used to programmatically submit a form (or
+ * some arbitrary data) to the server.
+ */
+export function useSubmit(): SubmitFunction {
+  return useSubmitImpl();
+}
+
+function useSubmitImpl(fetcherKey?: string): SubmitFunction {
+  let router = React.useContext(UNSAFE_DataRouterContext);
+  let defaultAction = useFormAction();
+
+  return React.useCallback(
+    (target, options = {}) => {
+      invariant(
+        router != null,
+        "useSubmit() must be used within a <DataRouter>"
+      );
+
+      if (typeof document === "undefined") {
+        throw new Error(
+          "You are calling submit during the server render. " +
+            "Try calling submit within a `useEffect` or callback instead."
+        );
+      }
+
+      let { method, encType, formData, url } = getFormSubmissionInfo(
+        target,
+        defaultAction,
+        options
+      );
+
+      let href = url.pathname + url.search;
+      let opts = {
+        replace: options.replace,
+        formData,
+        formMethod: method as FormMethod,
+        formEncType: encType as FormEncType,
+      };
+      if (fetcherKey) {
+        router.fetch(fetcherKey, href, opts);
+      } else {
+        router.navigate(href, opts);
+      }
+    },
+    [defaultAction, router, fetcherKey]
+  );
+}
+
+export function useFormAction(action = "."): string {
+  let routeContext = React.useContext(UNSAFE_RouteContext);
+  invariant(routeContext, "useLoaderData must be used inside a RouteContext");
```

Comment:
```suggestion
  invariant(routeContext, "useFormAction must be used inside a RouteContext");
```

Drive by comment while looking at the new goodies 🚗

Reply:
Thanks!  We've rebased this branch so I fixed this over in the new branch 👍